### PR TITLE
[Enhance] Remove CUDA part from diff_iou_rotated

### LINKED ISSUE #1922: wrong result of differentiable 3D IoU, occurring larger than 1
Thanks for reporting the unexpected results and we appreciate it a lot.

**Checklist**

1. I have searched related issues but cannot get the expected help.
2. I have read the [FAQ documentation](https://mmcv.readthedocs.io/en/latest/trouble_shooting.html) but cannot get the expected help.
3. The unexpected results still exist in the latest version.

**Describe the Issue**
A clear and concise description of what the bug is, including what results are expected and what the real results you got.

**Reproduction**

1. What command, code, or script did you run?

```bash
A placeholder for the command.
```

2. Did you make any modifications on the code? Did you understand what you have modified?

**Environment**

1. Please run `python -c "from mmcv.utils import collect_env; print(collect_env())"` to collect necessary environment information and paste it here.
2. You may add addition that may be helpful for locating the problem, such as
    - How you installed PyTorch [e.g., pip, conda, source]
    - Other environment variables that may be related (such as `$PATH`, `$LD_LIBRARY_PATH`, `$PYTHONPATH`, etc.)

**Error traceback**
If applicable, paste the error traceback here.

```none
A placeholder for traceback.
```

**Bug fix**
If you have already identified the reason, you can provide the information here. If you are willing to create a PR to fix it, please also leave a comment here and that would be much appreciated!

### PR DESCRIPTION
**Pull request by filaPro**
## Motivation

Fix #1922. Following [lilanxiao/Rotated_IoU#39#issuecomment-1146352088](https://github.com/lilanxiao/Rotated_IoU/issues/39#issuecomment-1146352088) we remove `cuda` and `cpp` part of `diff_iou_rotated` in favour of pure `pytorch` implementation. It has negligible affect on speed however the overall accuracy in corner cases is much better.

## Modification

Replace sorting vertices from `cuda` to `pytorch`. Split test to GPU and CPU.

## BC-breaking (Optional)

No, the public API is not changed. However we may want to move it from `mmcv.ops` as it does not require `cuda` now. `mmdet3d` does not use it in `master`. I've checked this PR with my FCAF3D PR [mmdetection3d#1547](https://github.com/open-mmlab/mmdetection3d/pull/1547). Also `diff_iou_rotated` is used in `mmrotate` `master`. Maybe @ZwwWayne or @Tai-Wang can have a look on `mmdetection3d` connection and @zytx121 for `mmrotate`.

### CODE REVIEW

**Comment by zhouzaida** [comment]
Hi @filaPro , thanks for your contributions. Could you add some benchmark results in the PR description? You can refer to https://github.com/open-mmlab/mmcv/pull/1718#issuecomment-1061531123.

**Comment by filaPro** [comment]
Hi @zhouzaida , done. I haven't tried `torch.cuda.Event` before, hope my script is right. Also don't think that the decrease of speed is important here.

File: tests/test_ops/test_diff_iou_rotated.py [hunk 1]
```diff
@@ -5,45 +5,57 @@
...
-        [[[0.5, 0.5, 1., 1., .0], [0.5, 0.5, 1., 1., np.pi / 2],
-          [0.5, 0.5, 1., 1., np.pi / 4], [1., 1., 1., 1., .0],
-          [1.5, 1.5, 1., 1., .0]]],
-        dtype=np.float32)
-
-    boxes1 = torch.from_numpy(np_boxes1).cuda()
-    boxes2 = torch.from_numpy(np_boxes2).cuda()
-
-    np_expect_ious = np.asarray([[1., 1., .7071, 1 / 7, .0]])
+def test_diff_iou_rotated_2d_gpu():
```
Comment (grimoire) [inline tests/test_ops/test_diff_iou_rotated.py:34]: As you have remove the cuda op, I think we does not need a test for cuda now.

File: mmcv/ops/diff_iou_rotated.py [hunk 2]
```diff
@@ -170,37 +146,121 @@ def sort_indices(vertices: Tensor, mask: Tensor) -> Tensor:
...
+    """Pad mask of invalid vertices with zeros.
+
+    Args:
+        num_valid (Tensor): (B, N) Number of valid vertices per pair of boxes.
+
+    Returns:
+       Tensor: (B, N, 9) Mask.
+    """
+    B, N = num_valid.size()
+    arange = torch.arange(9).unsqueeze(0).unsqueeze(0).repeat(B, N, 1).to(
```
Comment (grimoire) [inline mmcv/ops/diff_iou_rotated.py:260]: The repeat can be removed. `<` is a binary op, it can be broadcasted.

**Review by grimoire — APPROVED** [review]
LGTM

File: mmcv/ops/diff_iou_rotated.py [hunk 3]
```diff
@@ -170,37 +146,120 @@ def sort_indices(vertices: Tensor, mask: Tensor) -> Tensor:
...
+    ang = torch.atan2(y, x)
+    index = torch.argsort(ang, dim=-1)  # (B, N, 24)
+
+    # duplicate the first
+    temp = index[..., :1].clone()  # (B, N, 1)
+    index.scatter_(
+        dim=-1,
+        index=num_valid.unsqueeze(-1),
+        src=temp.expand(-1, -1, index.size(-1)))
+    return index[..., :9]
```
Comment (Tai-Wang) [inline mmcv/ops/diff_iou_rotated.py:246]: May need some comments for `:9`

File: mmcv/ops/diff_iou_rotated.py [hunk 4]
```diff
@@ -170,37 +146,120 @@ def sort_indices(vertices: Tensor, mask: Tensor) -> Tensor:
...
+
+@torch.no_grad()
+def generate_mask(num_valid: Tensor) -> Tensor:
+    """Pad mask of invalid vertices with zeros.
+
+    Args:
+        num_valid (Tensor): (B, N) Number of valid vertices per pair of boxes.
+
+    Returns:
+       Tensor: (B, N, 9) Mask.
```
Comment (Tai-Wang) [inline mmcv/ops/diff_iou_rotated.py:257]: Inaccurate indents

**Comment by Tai-Wang** [comment]
Only need to fix minor comments. Others LGTM.

**Comment by Tai-Wang** [comment]
@filaPro Great contribution! BTW, I am curious about the iou computation used in mmdet3d. Sometimes I feel like there are also some corner cases that BEV IoU can not be precisely computed, for example, for two strictly overlapped boxes on KITTI. Do you have any idea about it and is it related to similar problems as shown in this PR?

**Review by zytx121 — APPROVED** [review]
Nice work! LGTM.

**Review by Tai-Wang — APPROVED** [review]

**Comment by Tai-Wang** [comment]
LGTM. Just need to resolve conflicts

**Comment by zhouzaida** [comment]
Hi @filaPro , this PR can be merged after resolving conflicts.

**Comment by zytx121** [comment]
According to https://github.com/open-mmlab/mmcv/issues/2335, even if the two boxes overlap, there is still a gradient. Does this meet our expectations?
